### PR TITLE
Unify entity spawn packet sending

### PR DESCRIPTION
**Pull request by tigerw**
at the behest of xoft https://github.com/cuberite/cuberite/pull/4683#issuecomment-616217686

### CODE REVIEW

**Comment by madmaxoft** [comment]
I'd have preferred if there was no `SendSpawnEntityData`, but instead the `SendSpawnEntity` would have the entity-type forks that are currently in `GetProtocolSendSpawnEntityData`, and instead of `return {some, data}` there'd be a `SendSpawnObjectPacket(some, data)` call, or a call to specialized (overridable) `SendSpawnArrowEntity` etc.

File: src/Protocol/Protocol_1_8.cpp [hunk 1]
```diff
@@ -3918,3 +3850,46 @@ void cProtocol_1_8_0::WriteEntityProperties(cPacketizer & a_Pkt, const cEntity &
...
+				case PType::pkEgg: return 62;
+				case PType::pkGhastFireball: return 63;
+				case PType::pkFireCharge: return 64;
+				case PType::pkEnderPearl: return 65;
+				case PType::pkExpBottle: return 75;
+				case PType::pkSplashPotion: return 73;
+				case PType::pkFirework: return 76;
+				case PType::pkWitherSkull: return 66;
+				default: UNREACHABLE("Spawned an unhandled projectile");
+			}
```
Comment (peterbell10) [inline src/Protocol/Protocol_1_8.cpp:3888]: The idea is for the compiler to give you warnings if you don't cover all the enum cases. That only happens if there isn't a `default` label.
```suggestion
			}
			UNREACHABLE("Spawned an unhandled projectile");
```

Comment (peterbell10) [reply src/Protocol/Protocol_1_8.cpp:3888]: (This happens in a few other places as well)

File: src/Protocol/Protocol_1_8.h [hunk 2]
```diff
@@ -104,10 +103,9 @@ class cProtocol_1_8_0:
 	virtual void SendSetTitle                   (const cCompositeChat & a_Title) override;
 	virtual void SendSetRawTitle                (const AString & a_Title) override;
 	virtual void SendSoundParticleEffect        (const EffectID a_EffectID, int a_SrcX, int a_SrcY, int a_SrcZ, int a_Data) override;
-	virtual void SendSpawnFallingBlock          (const cFallingBlock & a_FallingBlock) override;
+	virtual void SendSpawnEntity                (const cEntity & a_Entity) override;
+	virtual void SendSpawnEntity                (const cEntity & a_Entity, const UInt8 a_ObjectType, const Int32 a_ObjectData);
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_8.h:107]: This should probably be a `protected` member, since it's expected to be called only from the inside of the protocol class.

File: src/Items/ItemItemFrame.h [hunk 3]
```diff
@@ -34,7 +34,6 @@ class cItemItemFrameHandler :
 
 		AddFaceDirection(a_BlockX, a_BlockY, a_BlockZ, a_BlockFace);  // Make sure block that will be occupied is free
 		BLOCKTYPE Block = a_World->GetBlock(a_BlockX, a_BlockY, a_BlockZ);
-		AddFaceDirection(a_BlockX, a_BlockY, a_BlockZ, a_BlockFace, true);  // We want the clicked block, so go back again
```
Comment (madmaxoft) [inline src/Items/ItemItemFrame.h:37]: I see this is "countered" by the protocol code, but what about the data saved to the chunk, is it valid then?

**Comment by tigerw** [comment]
Thanks for your reviews, I'll get on to it soon.

**Comment by tigerw** [comment]
I have addressed your comments, what d'you think?

File: src/Items/ItemItemFrame.h (hunk 3 above)
Comment (tigerw) [reply src/Items/ItemItemFrame.h]: Wiki says Vanilla changed storage to the block it's in back in 1.8, the last one we support.

**Comment by tigerw** [comment]
Going to merge

File: src/Protocol/Protocol.h [hunk 4]
```diff
@@ -259,9 +256,6 @@ class cProtocol
 	/** Returns the protocol-specific packet ID given the protocol-agnostic packet enum. */
 	virtual UInt32 GetPacketID(ePacketType a_Packet) = 0;
 
-	/** Converts eMonsterType to protocol-specific mob types */
-	virtual UInt32 GetProtocolMobType(eMonsterType a_MobType) = 0;
```
Comment (peterbell10) [inline src/Protocol/Protocol.h:263]: Why was this moved?

Comment (tigerw) [reply src/Protocol/Protocol.h:263]: It was too high up and forced ProtocolRecognizser to have a dummy `return 0` implementation, now the root of the inheritance chain starts lower, from a concrete protocol 1.8.

**Comment by tigerw** [comment]
Thanks!
